app/controllers/open_humans_profiles_controller.rb

```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+
+class OpenHumansProfilesController < ApplicationController
+  before_filter :require_user, except: [:index]
+
+  def start_auth
+    @user = current_user
+    if @user.open_humans_profile == nil
+      @user.open_humans_profile = OpenHumansProfile.new
+      @user.save
+    end
+    @oh_profile = @user.open_humans_profile
+
+    @client_id = ENV.fetch('OH_client_id')
+    redirect_to "https://www.openhumans.org/direct-sharing/projects/oauth2/authorize/?client_id=#{@client_id}&response_type=code"
+  end
+
+  def authorize
+    @user = current_user
+    @oh_profile = @user.open_humans_profile
+    @code = params[:code]
+
+    token_results = get_token(@code)
+    @oh_profile.expires_in = Time.now + token_results["expires_in"]
+    @oh_profile.access_token = token_results["access_token"]
+    @oh_profile.refresh_token = token_results["refresh_token"]
+
+    user_ids = get_open_humans_ids(@oh_profile.access_token)
+    @oh_profile.project_member_id = user_ids["project_member_id"]
+    @oh_profile.open_humans_user_id = user_ids["username"]
+    @oh_profile.save
+    upload_opensnp_id
+
+  end
```

Comment:
You don't need instance variables if you don't use them outside of the method. Also, since this is OAuth, why not use one of the numerous OAuth client libs instead of hand-rolling it? Also... 😉 It's good practice to put most of the logic of, like, everything in [service classes](https://hackernoon.com/service-objects-in-ruby-on-rails-and-you-79ca8a1c946e) instead of stuffing the controllers like a christmas goose. 🦆

Reply:
Do you have a recommendation for an OAuth client lib? The ones I saw while googling all felt a bit weird to me somehow, that's why I ended up writing it myself. 

But the service classes are a good idea. I'll refactor accordingly 👍

Reply:
I haven't looked into it for a while, so I'm not sure. It just seems like re-inventing the wheel.